velox/type/DecimalUtil.h

```diff
@@ -203,32 +204,57 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
-  /// Rescales a double value to decimal value of given precision and scale. The
-  /// output is rescaled value of int128_t or int64_t type. Returns error status
-  /// if fails.
-  template <typename TOutput>
-  inline static Status
-  rescaleDouble(double value, int precision, int scale, TOutput& output) {
+  /// Rescales a floating point value to decimal value of given precision and
+  /// scale. The output is rescaled value of int128_t or int64_t type. Returns
+  /// error status if fails.
+  template <typename TIntput, typename TOutput>
+  inline static Status rescaleFloatingPoint(
+      TIntput value,
+      int precision,
+      int scale,
+      TOutput& output) {
     if (!std::isfinite(value)) {
       return Status::UserError("The input value should be finite.");
     }
 
-    long double rounded;
-    // A double provides 16(±1) decimal digits, so at least 15 digits are
-    // precise.
-    if (scale > 15) {
-      // Convert value to long double type, as double * int128_t returns
-      // int128_t and fractional digits are lost. No need to consider 'toValue'
-      // becoming infinite as DOUBLE_MAX * 10^38 < LONG_DOUBLE_MAX.
-      const auto toValue = (long double)value * DecimalUtil::kPowersOfTen[15];
-      rounded = std::round(toValue) * DecimalUtil::kPowersOfTen[scale - 15];
+    uint8_t digits;
+    if constexpr (std::is_same_v<TIntput, float>) {
+      // A float provides between 6 and 7 decimal digits, so at least 6 digits
+      // are precise.
+      digits = 6;
+    } else {
+      // A double provides from 15 to 17 decimal digits, so at least 15 digits
+      // are precise.
+      digits = 15;
+      if (value <= std::numeric_limits<int128_t>::min() ||
```

Comment:
Why do we need this check? I assume we'll be counting number of digits before the period and we can check then, no?

Reply:
You're right. Updated this PR to check for lowest and max before casting to uint128_t for counting digits.